validator_client/slashing_protection/src/slashing_protection.rs

```diff
@@ -0,0 +1,378 @@
+}
+
+impl SafeFromSlashing<SignedAttestation> for HistoryInfo<SignedAttestation> {
+    type U = AttestationData;
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &AttestationData,
+        data_history: &[SignedAttestation],
+    ) -> Result<Safe, NotSafe> {
+        check_for_attester_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = self.verify_and_get_index(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                ValidityReason::SameVote => Ok(()),
+                _ => self
+                    .insert_and_write(SignedAttestation::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+impl SafeFromSlashing<SignedBlock> for HistoryInfo<SignedBlock> {
+    type U = BeaconBlockHeader;
+
+    fn verify_and_get_index(
+        &self,
+        incoming_data: &BeaconBlockHeader,
+        data_history: &[SignedBlock],
+    ) -> Result<Safe, NotSafe> {
+        check_for_proposer_slashing(incoming_data, data_history)
+    }
+
+    fn update_if_valid(&mut self, incoming_data: &Self::U) -> Result<(), NotSafe> {
+        let check = self.verify_and_get_index(incoming_data, &self.data[..]);
+        match check {
+            Ok(safe) => match safe.reason {
+                // Casting the same vote, no need to add it go the history
+                ValidityReason::SameVote => Ok(()),
+                // New attestation, add it to memory and file history
+                _ => self
+                    .insert_and_write(SignedBlock::from(incoming_data), safe.insert_index)
+                    .map_err(|e| NotSafe::IOError(e.kind())),
+            },
+            Err(notsafe) => Err(notsafe),
+        }
+    }
+}
+
+/// Struct used for checking if attestations or blockheader are safe from slashing.
+#[derive(Debug)]
+pub struct HistoryInfo<T: Encode + Decode + Clone + PartialEq> {
+    filepath: PathBuf,
+    data: Vec<T>,
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> PartialEq for HistoryInfo<T> {
+    fn eq(&self, other: &Self) -> bool {
+        let my_data = &self.data;
+        let other_data = &other.data;
+        self.filepath == other.filepath && *my_data == *other_data
+    }
+}
+
+impl<T: Encode + Decode + Clone + PartialEq> HistoryInfo<T> {
+    /// Inserts the incoming data in the in-memory history, and writes it to the history file.
+    pub fn insert_and_write(&mut self, data: T, index: usize) -> IOResult<()> {
+        self.data.insert(index, data);
+
+        // Creating file if it doesn't exist, else opening it.
+        let mut file = File::create(self.filepath.as_path())?;
+
+        // Locking the file to make sure we're atomically writing to it
+        file.lock_exclusive()?;
```

Comment:
According to the [docs](https://docs.rs/fs2/0.4.3/fs2/trait.FileExt.html#tymethod.lock_exclusive), this will block the current thread until it can obtain a lock on the file.

If there is contention for the file then one of those process will just block here. This seems quite reasonable in the best-case; each process won't have to block for long. However, if one of the processes crashes between unlocking and locking the file (e.g., crash, power failure), then future threads may just block here quietly forever. It's worth noting that this is a OS-level lock and may be triggered from outside of our process or outlive it.

For our application, I think using [`try_lock_exclusive`](https://docs.rs/fs2/0.4.3/fs2/trait.FileExt.html#tymethod.try_lock_exclusive) and immediately returning an error is ideal. It means we'll push the "retry" logic back up into the validator client, but I think that's fine. There we can use timer `futures` to keep polling a locked file, potentially logging warnings if we detect contention.

Also, I think there's a race condition here because we lock the file when `verifying`, instead of locking the file for the entire lifetime of `HistoryInfo` (i.e., taking the lock in the constructor and unlocking it on `Drop`). Consider two processes both accessing the same file:

1. Process `A` opens an existing `HistoryInfo`, loads `n` previous votes into memory.
1. Process `B` does the same.
1. Process `A` calls `update_if_valid(v_1)`.
1. Process `B` calls `update_if_valid(v_2)`.
1. `v_1` is not in the history.